taichi/transforms/lower_ast.cpp

```diff
@@ -167,11 +167,50 @@ class LowerAST : public IRVisitor {
       auto end = stmt->end;
       begin->flatten(flattened);
       end->flatten(flattened);
-      auto &&new_for = std::make_unique<RangeForStmt>(
-          stmt->parent->lookup_var(stmt->loop_var_id[0]), begin->stmt,
-          end->stmt, std::move(stmt->body), stmt->vectorize, stmt->parallelize,
-          stmt->block_dim, stmt->strictly_serialized);
-      flattened.push_back(std::move(new_for));
+      static bool is_good_range_for = true; // TODO(archibate): detect if outer-most one
+      if (1 || is_good_range_for) { // #578
+        is_good_range_for = false;
+        auto &&new_for = std::make_unique<RangeForStmt>(
+            stmt->parent->lookup_var(stmt->loop_var_id[0]), begin->stmt,
+            end->stmt, std::move(stmt->body), stmt->vectorize, stmt->parallelize,
+            stmt->block_dim, stmt->strictly_serialized);
+        flattened.push_back(std::move(new_for));
+      } else {
+        // transform into a structure as
+        // while (1) { cond; if (no active) break; original body...}
+        //auto loop_var = Stmt::make<AllocaStmt>(DataType::i32);
+        auto loop_var = stmt->parent->lookup_var(stmt->loop_var_id[0]);
+        flattened.push_back(Stmt::make<LocalStoreStmt>(loop_var, begin->stmt));
+        auto loop_var_addr = LaneAttribute<LocalAddress>(
+            LocalAddress(loop_var->as<AllocaStmt>(), 0));
+        // auto loop_var_store = Stmt::make<LocalStoreStmt>(loop_var_addr, begin->stmt);
+        auto loop_var_load = Stmt::make<LocalLoadStmt>(loop_var_addr);
+        auto cond_stmt = Stmt::make<BinaryOpStmt>(BinaryOpType::cmp_lt,
+            std::move(loop_var_load).get(), end->stmt);
+
+        auto &&new_while = std::make_unique<WhileStmt>(std::move(stmt->body));
+        auto mask = std::make_unique<AllocaStmt>(DataType::i32);
+        new_while->mask = mask.get();
+        auto &stmts = new_while->body;
+        for (int i = 0; i < (int)flattened.size(); i++) {
+          stmts->insert(std::move(flattened[i]), i);
+        }
+        // insert break
+        stmts->insert(
+            std::make_unique<WhileControlStmt>(new_while->mask, std::move(cond_stmt).get()),
```

Comment:
How does `WhileControlStmt` control `while` by `mask`?

Reply:
`mask` is for vectorization on CPUs. It indicates which lanes of a SIMD vector are active.

Reply:
So it's `(mask & 1) == cond` on single CPU?

Reply:
> WhileControlStmt will just set the lane masks to 0 so that inactive lanes (i.e. those already exited) will not get executed.

Cool impl! will vectorized ones do atomic read mask to become zero only if mask bit nr > current bit nr?